### ISSUE TITLE
Convenience API for downloading files of MemFS

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#2104]
## 🚀 Feature

A convenience API to download files out of the memory file system. Possibly we could also have an API to zip a directory and download the zip file.

### Motivation

We've made it a lot easier to get files into the Emscripten file system, but we still don't have any designated way to get files back out. This would probably be helpful for people. See for instance #2103. We've gotten a decent number of questions about this over time. 

One question is how best to deal with Node.

**Comment by ryanking13** [comment]
I agree that we need some more documentation in using the filesystem. It's very common to write a file into a virtual file system or retrieve a file from it.

But I'm not sure whether we need an API for downloading a file.
I think our job is to provide an easy way (or clear documentation) to retrieve a file from the virtual filesystem to a JavaScript object. After that, people can easily use third-party libraries that can deal with downstream tasks like downloading/uploading/etc (e.g. [file-saver](https://www.npmjs.com/package/file-saver))

**Comment by hoodmane** [comment]
I think we did this, it's called `pyodide.http.pyfetch`/

**Comment by ryanking13** [comment]
> I think we did this, it's called `pyodide.http.pyfetch`.

No, this issue is about downloading a file from a browser virtual file system to a device.

**Comment by hoodmane** [comment]
Ahh sorry! 

So part of my motivation for this is imagining a workflow with `micropip` where you use it to "install" packages once, then use `micropip.freeze` to generate a `packages.json` file that you can then deploy as part of your application and then you won't need to wait for micropip to download dependencies. This seems like it might be annoying to use without a good way of downloading the generated json.

**Comment by ryanking13** [comment]
> So part of my motivation for this is imagining a workflow with `micropip` where you use it to "install" packages once, then use `micropip.freeze` to generate a `packages.json` file that you can then deploy as part of your application and then you won't need to wait for micropip to download dependencies.

Yeah, I like the idea of freezing dependencies.

I guess the freeze API would be look like:

```js
let packageJson = pyodide.runPython("micropip.freeze()")
pyodide.download(packageJson, "packages.json")
// or
let packageJson = pyodide.freeze()
pyodide.download(packageJson, "packages.json")
// or
pyodide.freeze("packages.json")
```

In that case, I think the download API should be "downloading JavaScript object" not "downloading a file in virtual file system", because one can easily do:

```js
let file = pyodide.FS.readFile("some_file.ext")
pyodide.download(file, "some_file.ext")
```

**Comment by hoodmane** [comment]
> In that case, I think the download API should be "downloading JavaScript object" 

Right. I think I said file system API because it made it sound more closely attached to other stuff. Download a JS object sounds a bit like it has no relationship to the core purpose of Pyodide, it's just a utility. Not that that means we can't provide it.

**Comment by ryanking13** [comment]
> Download a JS object sounds a bit like it has no relationship to the core purpose of Pyodide, it's just a utility. Not that that means we can't provide it.

Yes, that's my point. There are other utility libraries which does that functionality very well. So I'm not sure we should embed it in Pyodide or tell people to use other libraries.

If we think downloading a JavaScript object in Pyodide will happen really often, it would be great to add that feature. Maybe we can discuss this after designing freezing dependencies feature.
